File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.1_RabbitMQ/02_exchanges_routing/topic_producer.py
#!/usr/bin/env python3
"""
RabbitMQ Topic Exchange Producer

Routes messages based on routing key patterns.
Use case: Logging systems, event distribution with filtering
"""

import pika
import json
import sys
from datetime import datetime


def create_connection():
    """Create connection to RabbitMQ"""
    credentials = pika.PlainCredentials('admin', 'admin123')
    parameters = pika.ConnectionParameters(
        host='localhost',
        port=5672,
        credentials=credentials
    )
    return pika.BlockingConnection(parameters)


def publish_message(channel, exchange, routing_key, message):
    """Publish message to topic exchange"""
    channel.basic_publish(
        exchange=exchange,
        routing_key=routing_key,
        body=json.dumps(message),
        properties=pika.BasicProperties(
            delivery_mode=2,
            content_type='application/json'
        )
    )
    print(f"✓ Sent [{routing_key}]: {message['text']}")


def main():
    """Main function"""
    exchange_name = 'topic_logs'

    # Connect to RabbitMQ
    connection = create_connection()
    channel = connection.channel()

    # Declare topic exchange
    channel.exchange_declare(
        exchange=exchange_name,
        exchange_type='topic',
        durable=True
    )
    print(f"✓ Exchange '{exchange_name}' declared (type: topic)\n")

    # Send messages with hierarchical routing keys
    # Format: <facility>.<severity>.<component>
    messages = [
        {
            'routing_key': 'app.info.startup',
            'message': {
                'text': 'Application started',
                'facility': 'app',
                'severity': 'info',
                'component': 'startup',
                'timestamp': datetime.now().isoformat()
            }
        },
        {
            'routing_key': 'app.error.database',
            'message': {
                'text': 'Database connection failed',
                'facility': 'app',
                'severity': 'error',
                'component': 'database',
                'timestamp': datetime.now().isoformat()
            }
        },
        {
            'routing_key': 'kernel.critical.memory',
            'message': {
                'text': 'Out of memory',
                'facility': 'kernel',
                'severity': 'critical',
                'component': 'memory',
                'timestamp': datetime.now().isoformat()
            }
        },
        {
            'routing_key': 'app.warning.authentication',
            'message': {
                'text': 'Multiple failed login attempts',
                'facility': 'app',
                'severity': 'warning',
                'component': 'authentication',
                'timestamp': datetime.now().isoformat()
            }
        },
        {
            'routing_key': 'kernel.info.network',
            'message': {
                'text': 'Network interface up',
                'facility': 'kernel',
                'severity': 'info',
                'component': 'network',
                'timestamp': datetime.now().isoformat()
            }
        }
    ]

    for item in messages:
        publish_message(
            channel,
            exchange_name,
            item['routing_key'],
            item['message']
        )

    connection.close()
    print(f"\n📊 Sent {len(messages)} messages to '{exchange_name}' exchange")


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('\n✗ Interrupted')
        sys.exit(0)
    except Exception as e:
        print(f'✗ Error: {e}')
        sys.exit(1)

File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.1_RabbitMQ/02_exchanges/topic_producer.ts
#!/usr/bin/env ts-node
import * as amqp from 'amqplib';
import { Connection, Channel } from 'amqplib';

/**
 * Topic message interface
 */
interface TopicMessage {
    routing_key: string;
    message: string;
    timestamp: string;
}

/**
 * Send message to topic exchange with pattern-based routing key
 */
async function sendTopicMessage(routingKey: string, messageText: string): Promise<void> {
    let connection: Connection | null = null;
    let channel: Channel | null = null;

    try {
        const connectionUrl = 'amqp://admin:password@localhost:5672/';
        connection = await amqp.connect(connectionUrl);
        channel = await connection.createChannel();

        // Declare topic exchange
        const exchangeName = 'logs_topic';
        await channel.assertExchange(exchangeName, 'topic', {
            durable: true
        });

        const message: TopicMessage = {
            routing_key: routingKey,
            message: messageText,
            timestamp: new Date().toISOString()
        };

        channel.publish(
            exchangeName,
            routingKey,
            Buffer.from(JSON.stringify(message)),
            {
                persistent: true,
                contentType: 'application/json'
            }
        );

        console.log(`✓ Sent [${routingKey}]: ${messageText}`);

    } catch (error) {
        console.error('Error sending message:', error);
        throw error;
    } finally {
        try {
            if (channel) await channel.close();
            if (connection) await connection.close();
        } catch (closeError) {
            console.error('Error closing connection:', closeError);
        }
    }
}

// Main execution
if (require.main === module) {
    const args = process.argv.slice(2);

    if (args.length < 2) {
        console.log('Usage: ts-node topic_producer.ts <routing.key> <message>');
        console.log('Examples:');
        console.log('  ts-node topic_producer.ts kern.critical "System crash"');
        console.log('  ts-node topic_producer.ts app.error "Database error"');
        console.log('  ts-node topic_producer.ts app.info "User logged in"');
        process.exit(1);
    }

    const routingKey = args[0];
    const message = args.slice(1).join(' ');

    sendTopicMessage(routingKey, message).catch(error => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}

export { sendTopicMessage, TopicMessage };
